### PR TITLE
Fix blank screen is shown when opening a report screen after relaunching app

### PR DESCRIPTION
**Pull request by bernhardoj**
### Details

The `withOnyx` returns `null` when the data is still loading. We already put some `initialValue` to the required onyx data, but Onyx has a bug that doesn't accept falsey `initialValue`. This PR also fixes some App issue that arise after fixing the onyx bug.

### Fixed Issues

$ https://github.com/Expensify/App/issues/35906
PROPOSAL: https://github.com/Expensify/App/issues/35906#issuecomment-1930326277

### Tests

Same as QA Steps

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA Steps

### QA Steps

A.
1. Close the app if it's opened
2. Reopen the app
3. Open any chat
4. Do step 3 on several chats
5. Verify the chat screen/skeleton is shown instead of blank screen for a brief moment

B.
1. Open a workspace chat
2. Turn off internet connection
3. Request money from the workspace
5. Open the transaction thread
6. Verify you are able to see the transaction thread details instead of infinite loading

C.
1. Create a new workspace room
2. Verify you are navigated to the workspace room chat and not to concierge chat
3. Verify the room name is correct

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/50919443/1d86a164-b6b5-4210-8d9f-14607ade69c7

https://github.com/Expensify/App/assets/50919443/223dff13-3ca5-4b58-9b54-c9d43b995b2b

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/react-native-onyx/assets/50919443/5ed7b31a-59f4-4117-99c6-2d74d45a417b

https://github.com/Expensify/App/assets/50919443/787d2859-af3f-4208-a039-864af55129bf

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/50919443/6c86f3c6-8c23-42c0-9ab9-040c38062d46

https://github.com/Expensify/App/assets/50919443/b066c86e-876d-4332-9155-28e893329b1f

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/50919443/17de4e90-404d-4f85-917f-9047408448f5

https://github.com/Expensify/App/assets/50919443/f6bc4e0a-d2c5-4cc2-9c11-af10c2e91541

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/50919443/f12bd94e-137a-4ecc-a051-86f353535d01

https://github.com/Expensify/App/assets/50919443/f509d9b3-7cd0-48e3-af90-902463cd121a

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/react-native-onyx/assets/50919443/03d7d07c-8d1a-4c6a-9fa4-e479bec57ce2

https://github.com/Expensify/App/assets/50919443/fccfa80d-a20b-444d-9f64-344153d9b91c

</details>

### CODE REVIEW

File: src/libs/actions/Report.ts [hunk 1]
```diff
@@ -625,13 +625,6 @@ function openReport(
         parameters.shouldRetry = false;
     }
 
-    const report = ReportUtils.getReport(reportID);
-    // If we open an exist report, but it is not present in Onyx yet, we should change the method to set for this report
-    // and we need data to be available when we navigate to the chat page
-    if (isEmptyObject(report)) {
-        optimisticData[0].onyxMethod = Onyx.METHOD.SET;
-    }
-
```
Comment (bernhardoj) [inline src/libs/actions/Report.ts:634]: This is to fix all the found issues from our PR 
https://github.com/Expensify/App/issues/37247, https://github.com/Expensify/App/issues/37234, https://github.com/Expensify/App/issues/37300, and https://github.com/Expensify/App/issues/37251.

The root cause is explained [here](https://github.com/Expensify/App/issues/37247#issuecomment-1965912623)

**Comment by bernhardoj** [comment]
There are 2 issues left however and they have the same issue too.
1. Chat Report is shown as the report name when creating a new workspace room
image

The report name is overwritten with the default name here because the onyx data is not ready yet.
https://github.com/Expensify/App/blob/39a257d1aa850e875bcf604f6c7b039527234952/src/libs/actions/Report.ts#L572-L576

2. OpenReport is still called even though we have a comment here mentioning that we use SET so the report data is available immediately
https://github.com/Expensify/App/blob/39a257d1aa850e875bcf604f6c7b039527234952/src/libs/actions/Report.ts#L1767-L1770

so we can avoid calling OpenReport.
https://github.com/Expensify/App/blob/39a257d1aa850e875bcf604f6c7b039527234952/src/pages/home/ReportScreen.js#L355-L362

However, the report data is not available immediately.

The `optimisticData` is updated through `Onyx.update` calls which use promises that delay the operation.
https://github.com/Expensify/react-native-onyx/blob/034aa0e7858ec4f8a859117b5f527c793b0a61f7/lib/Onyx.js#L1494-L1518


We can fix this using `Onyx.set` or `Onyx.multiSet` directly, so the report (and other required data) is 100% immediately available. So far, I have used this in `Report.openReport` and `Report.addPolicyReport`, and it works well. What do you think?

**Comment by bernhardoj** [comment]
Or perhaps we should remove the `allowStaleData` because the component expects the report data to be available (not pending to set/merge) when mounted. So far I can't notice the blank screen after removing it and returning undefined from `reportWithoutHasDraftSelector` instead of null.

**Comment by bernhardoj** [comment]
Hmm, the only idea left I have is to replace `withOnyx` with `useOnyx` hook.

**Comment by bernhardoj** [comment]
Based on the video, looks like the report screen is rendered immediately as shown by the (broken) composer, I will check it too tomorrow to see if I can reproduce and find something about it.

**Comment by bernhardoj** [comment]
> This line of code was introduced here, and I'm also testing to remove it.

I think we can keep it like that. To solve the name issue in my comment [here](https://github.com/Expensify/App/pull/37596#issuecomment-1973088263), I update the code to call OpenReport only after the report onyx is loaded. I think we can convert only some of the onyx data to `useOnyx` to reduce the risk of regression. I have pushed the update so you can compare and test it.

**Comment by bernhardoj** [comment]
Ok I can reproduce the blank list. I checked the onyx data of ReportActionsView
https://github.com/Expensify/App/blob/7ef8e8c85d8c849f130ae0174de71e163aec7a0d/src/pages/home/report/ReportActionsView.tsx#L622-L633

and it's the `transactionThreadReport` that's not available yet, so I set the initial value to it.

**Review by ntdiary — CHANGES_REQUESTED** [review]

File: src/pages/home/ReportScreen.tsx [hunk 2]
```diff
@@ -489,6 +467,16 @@ function ReportScreen({
         // eslint-disable-next-line react-hooks/exhaustive-deps
     }, []);
 
+    useEffect(() => {
+        // Call OpenReport only if we are not linking to a message or the report is not available yet
+        if (reportResult.status === 'loading' || (reportActionIDFromRoute && report.reportID)) {
```
[the remark this answers is not in the document]
Comment (bernhardoj) [reply src/pages/home/ReportScreen.tsx]: Oh, nice one. Updated.

**Comment by bernhardoj** [comment]
Btw, I'm still considering whether to include the report onyx loading state to `isLoading` here

https://github.com/Expensify/App/blob/a03fecc80e5326a24644254f682441aa50585347/src/pages/home/ReportScreen.tsx#L393

I think we should add it. Do you agree?

**Comment by bernhardoj** [comment]
Thanks for the updated! Merged with main

**Review by ntdiary — APPROVED** [review]
LGTM. :)
Additionally, A and C are online cases, B is offline. And B, C are regression cases that were reported previously.

https://github.com/Expensify/App/assets/8579651/d31363f6-91cd-44a5-b951-605d78e8cc0c

https://github.com/Expensify/App/assets/8579651/ac0dc847-1d8a-413f-8ac0-9c902b0d412f

**Review by yuwenmemon — APPROVED** [review]

**Comment by yuwenmemon** [comment]
Thanks!

**Pull request merged**